### PR TITLE
SAK-29426 - LTI webservice that allows to add an LTI tool to sites

### PR DESCRIPTION
**Pull request by mcarropellicer**
SAK-29426 - LTI webservice that allows to add an LTI tool to sites

### CODE REVIEW

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            } else {
+                tool.setTitle(toolManager.getTool(toolid).getTitle());
+            }
+
+            siteService.save(siteEdit);
+            LOG.info("LTI tool added for site:" + siteId);
+
+            return "success";
+        } catch (Exception e) {
+            LOG.error("WS addToolAndPageToSite(): " + e.getClass().getName() + " : " + e.getMessage());
```
Comment (ottenhoff) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: Can you fix this error statement?

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+				if(prop != null && !"".equals(prop)) {
+					int index = prop.indexOf(":");
+					if(index >= 0) {
+						try {
+							String propName = prop.substring(0, index);
+							String propValue = prop.substring(index+1);
+							
+							if(propValue != null && !"".equals(propValue)) {
+								Properties propsedit = tool.getPlacementConfig();
+		                        propsedit.setProperty(propName, propValue);
```
Comment (ottenhoff) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: Can you fix tabs vs spaces here?

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+@WebService
+@SOAPBinding(style = SOAPBinding.Style.RPC, use = SOAPBinding.Use.LITERAL)
+public class SakaiLTI extends AbstractWebService {
+
+    private static final Log LOG = LogFactory.getLog(SakaiLTI.class);
+	
+	
+	@WebMethod(exclude = true)
+    public void init() {
```
Comment (ottenhoff) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: Can you fix tabs vs spaces here?

**Comment by ottenhoff** [comment]
Looks pretty good. This just needs some formatting help.  In vim, I use ":set list" to see tabs vs space characters.

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java (hunk 1 above)
Comment (mcarropellicer) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: I'm not the author of the patch so sorry for the mistake, will correct it and squash the changes in one commit.

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java (hunk 2 above)
Comment (mcarropellicer) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: Of course.

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java (hunk 3 above)
Comment (mcarropellicer) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: Of course!

**Comment by mcarropellicer** [comment]
Hi @ottenhoff, all suggestions implemented.

Thanks

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java (hunk 2 above)
Comment (ottenhoff) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: This tabs/spaces issue still needs fixing

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+			LOG.error("WS addLTITool(): " + e.getClass().getName() + " : " + e.getMessage());
+			e.printStackTrace();
+			return e.getClass().getName() + " : " + e.getMessage();
+		}
+	}
+	
+	private void setToolProperties(ToolConfiguration tool, String propList) {
+		if(propList != null) {
+			for(String prop : propList.split(",")) {
+				if(prop != null && !"".equals(prop)) {
```
Comment (ottenhoff) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: Can you use StringUtils.isNotEmpty(prop) instead?

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+					if(index >= 0) {
+						try {
+							String propName = prop.substring(0, index);
+							String propValue = prop.substring(index+1);
+							
+							if(propValue != null && !"".equals(propValue)) {
+								Properties propsedit = tool.getPlacementConfig();
+		                        propsedit.setProperty(propName, propValue);
+							}
+						} catch(Exception e){}
```
Comment (ottenhoff) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: Can you log this exception?

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java (hunk 2 above)
Comment (mcarropellicer) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: Resolved? Thanks

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java (hunk 4 above)
Comment (mcarropellicer) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: Done

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java (hunk 5 above)
Comment (mcarropellicer) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: Done

File: webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+		if(propList != null) {
+			for(String prop : propList.split(",")) {
+				if(prop != null && StringUtils.isNotEmpty(prop)) {
+					int index = prop.indexOf(":");
+					if(index >= 0) {
+						try {
+							String propName = prop.substring(0, index);
+							String propValue = prop.substring(index+1);
+							
+							if(propValue != null && StringUtils.isNotEmpty(propValue)) {
```
Comment (ottenhoff) [inline webservices/cxf/src/java/org/sakaiproject/webservices/SakaiLTI.java]: StringUtils methods are null safe .... so there is no need to check for null

**Comment by ottenhoff** [comment]
I don't think we should add a new LTI file .... let's just add this method to the existing SakaiScript.java

**Comment by ern** [comment]
Miguel,

This contributions author was Francisco Segura can you please update the commit.

Thanks

**Comment by mcarropellicer** [comment]
Done! Thanks.

**Pull request merged**